framework/src/modules/chain/bft/bft.js

```diff
@@ -14,273 +14,135 @@
-			: this.finalizedHeight;
+	async _initFinalityManager() {
+		// Check what finalized height was stored last time
+		const finalizedHeightStored =
+			parseInt(await this.ChainMetaEntity.getKey(KEYS.FINALIZED_HEIGHT)) || 0;
+
+		// Check BFT migration height
+		// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#backwards-compatibility
+		const bftMigrationHeight =
+			this.constants.startingHeight - this.constants.activeDelegates * 2;
+
+		// Choose max between stored finalized height or migration height
+		const finalizedHeight = Math.max(finalizedHeightStored, bftMigrationHeight);
+
+		// Initialize consensus manager
+		return new FinalityManager({
+			finalizedHeight,
+			activeDelegates: this.constants.activeDelegates,
+		});
 	}
 
 	/**
-	 * Use existing block headers and re-compute all information
+	 * Get the last block height from storage
+	 *
+	 * @return {Promise<number>}
+	 * @private
 	 */
-	recompute() {
-		this.state = {};
-		this.finalizedHeight = 0;
-		this.prevotedConfirmedHeight = 0;
-		this.preVotes = {};
-		this.preCommits = {};
-
-		this.headers.items.forEach(header => {
-			this.updatePreVotesPreCommits(header);
-			this.updatePreVotedAndFinalizedHeight();
-		});
+	async _getLastBlockHeight() {
+		const lastBlock = await this.BlockEntity.get(
+			{},
+			{ limit: 1, sort: 'height:desc' }
+		);
+		return lastBlock.length ? lastBlock[0].height : 0;
 	}
 
 	/**
-	 * Verify if the block header is good for current chain
+	 * Load blocks into consensus manager fetching from storage
 	 *
-	 * @param {BlockHeader} blockHeader
+	 * @param {integer} fromHeight - The start height to fetch and load
+	 * @param {integer} tillHeight - The end height to fetch and load
+	 * @return {Promise<void>}
 	 */
-	verifyBlockHeaders(blockHeader) {
-		// We need minimum processingThreshold to decide
-		// if prevotedConfirmedUptoHeight is correct
-		if (
-			this.headers.length >= this.processingThreshold &&
-			blockHeader.prevotedConfirmedUptoHeight !== this.prevotedConfirmedHeight
-		) {
-			throw new Error('Wrong prevotedConfirmedHeight in blockHeader.');
-		}
-
-		// Find top most block forged by same delegate
-		const delegateLastBlock = this.headers
-			.top(this.processingThreshold)
-			.reverse()
-			.find(
-				header => header.delegatePublicKey === blockHeader.delegatePublicKey
-			);
-
-		if (!delegateLastBlock) {
-			return true;
-		}
-
-		if (
-			delegateLastBlock.maxHeightPreviouslyForged ===
-				blockHeader.maxHeightPreviouslyForged &&
-			delegateLastBlock.height >= blockHeader.height
-		) {
-			// Violation of the fork choice rule as delegate moved to different chain
-			// without strictly larger maxHeightPreviouslyForged or larger height as
-			// justification. This in particular happens, if a delegate is double forging.
-			throw new Error(
-				'Violation of fork choice rule, delegate moved to a different chain'
-			);
-		}
+	async loadBlocks({ fromHeight, tillHeight }) {
```

Comment:
Shouldn't this function also be a private function?

Reply:
I kept it public if one need to load blocks to BFT, may be synchroniser or some other logic.